Option to send email to user when they are deactivated

juliaBichler01 opened the pull request:
Resolves https://github.com/zulip/zulip/issues/18943

This pull request adds the option to send a email to users after deactivating them.

![chrome-capture](https://user-images.githubusercontent.com/74348920/151515246-47bcee02-8c63-4f4e-b493-4f4d41c65e0e.gif)

![image](https://user-images.githubusercontent.com/74348920/151515300-2fba16c2-714d-4160-b555-987189281cba.png)

juliaBichler01 commented:
@alya i think this is looking pretty good now, so a review would be great!

alya commented:
Hm, I'm getting an error if I specify an email to send (things work fine if I leave the checkbox unchecked).

<img width="754" alt="Screen Shot 2021-12-06 at 9 20 46 AM" src="https://user-images.githubusercontent.com/2090066/144892373-6c343048-f4f7-4324-9eb7-840eb1619fd4.png">

Also, a couple of tweaks:
1) Let's change the checkbox text "Seen an email" -> "Send a notification email"
2) I think we need to add a header to the text box to make it more clear what's going on. Let's try:
"Message for {user name} ({user email}):"

Thanks!

alya commented:
We'll want to add a section to https://zulip.com/help/deactivate-or-reactivate-a-user documenting exactly what happens when you use this option.

juliaBichler01 commented:
@alya okay, the "Save failed" is interesting, because i am unfortunately not able to replicate it... Did you have any special settings while testing this or something like that?

But i will implement the other comments

alya commented:
I didn't specifically play with the settings... Perhaps try it with users with different permissions in the dev environment? I'm not sure.

juliaBichler01 commented:
I tried it again with all combinations of settings, account of the user that's doing the deactivation and account of the user that's getting deactivated that i could find, but i still couldn't replicate the problem.
Did this happen to you every time you tried it or just if Desdemona tries to deactivate Polonius?

juliaBichler01 commented:
@alya Do you maybe have any console output when you have this error? 
Because that would really help, as i am still not able to reproduce this

alya commented:
As Desdemona, deactivating Iago worked fine, but deactivating Zoe did not. Here's what my console is showing:

```
2021-12-14 06:06:08.084 ERR  [zerver.middleware.json_error_handler] Traceback (most recent call last):
  File "/srv/zulip/zerver/lib/send_email.py", line 121, in render_templates
    html_message = loader.render_to_string(template_prefix + ".html", context)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/template/loader.py", line 61, in render_to_string
    template = get_template(template_name, using=using)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/template/loader.py", line 19, in get_template
    raise TemplateDoesNotExist(template_name, chain=chain)
django.template.exceptions.TemplateDoesNotExist: zerver/emails/deactivate.html

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 33, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 169, in rest_dispatch
    return target_function(request, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 739, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 796, in _wrapped_view_func
    **kwargs,
  File "/srv/zulip/zerver/decorator.py", line 777, in authenticate_log_and_execute_json
    return limited_view_func(request, user_profile, *args, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 984, in wrapped_func
    return func(request, *args, **kwargs)
  File "/srv/zulip/zerver/lib/request.py", line 443, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/srv/zulip/zerver/views/users.py", line 112, in deactivate_user_backend
    return _deactivate_user_profile_backend(request, user_profile, target, email_text)
  File "/srv/zulip/zerver/views/users.py", line 141, in _deactivate_user_profile_backend
    context={"text": email_text},
  File "/srv/zulip/zerver/lib/send_email.py", line 241, in send_email
    realm=realm,
  File "/srv/zulip/zerver/lib/send_email.py", line 146, in build_email
    (html_message, message, email_subject) = render_templates()
  File "/srv/zulip/zerver/lib/send_email.py", line 126, in render_templates
    html_message = loader.render_to_string(compiled_template_prefix + ".html", context)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/template/loader.py", line 61, in render_to_string
    template = get_template(template_name, using=using)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/template/loader.py", line 19, in get_template
    raise TemplateDoesNotExist(template_name, chain=chain)
django.template.exceptions.TemplateDoesNotExist: zerver/emails/compiled/deactivate.html

2021-12-14 06:06:08.086 INFO [zr] 10.0.2.2        DELETE  500 119ms (db: 22ms/29q) /json/users/7 (9@zulip via Mozilla)
2021-12-14 06:06:08.086 ERR  [django.request] Internal Server Error: /json/users/7
2021-12-14 06:06:08.089 ERR  [django.server] "DELETE /json/users/7 HTTP/1.1" 500 49
```

juliaBichler01 commented:
Okay, i was able to recreate this error, but only after i deleted the deactivate.html file in the "compiled" folder. But then it didn't work all the time, not just for some users...
I guess you did provision or call "scripts/setup/inline_email_css.py", right?

alya commented:
Ah, sorry I missed your comment earlier. I ran `tools/provision` as usual.

alya commented:
Please let me know if/when this is ready for another review.

alya commented:
Nice! Could you add to the dialog box the same line breaks you added to the email?  (i.e. blank space after the greeting and between the paragraphs)

alya commented:
I have some minor changes I'd like to make to the help page, but I think this can happen whenever we're ready to merge the PR.

@timabbott could you review the code at this point?

juliaBichler01 commented:
@alya like this?

![tmp](https://user-images.githubusercontent.com/74348920/151514618-4edb79a2-798f-495e-91a6-b5156eceafd6.png)

(i pushed those changes already)

timabbott commented on static/js/settings_users.js:
```diff
@@ -445,7 +448,17 @@ function confirm_deactivation(row, user_id, status_field) {
+            data = {
+                email_text: $(".email_field_textarea").val(),
+                realm_name: page_params.realm_name,
+                username: user.full_name,
```
I don't think you should need the `realm_name` or `username` parameters in this HTTP request; the server already has these details for the target user.

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -1,7 +1,29 @@
     {{#tr}}
         When you deactivate <z-user></z-user>, they will be immediately logged out.
-        {{#*inline "z-user"}}<strong>{{username}}</strong>{{#if email}} &lt;{{email}}&gt;{{/if}}{{/inline}}
+        {{#*inline "z-user"}}<strong>{{username}}</strong>{{#if email}} &#40;{{email}}&#41;{{/if}}{{/inline}}
```
Why these changes?

juliaBichler01 replied:
I used round brackets when the user email was still displayed in the text of the checkbox, because i think that those make the next look better and then i adapted those, so they would look the same.
But i changed them back

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -1,7 +1,29 @@
+    <input type="checkbox" class="send_email" data-key="{{ key }}" />
+    {{#tr}}
+    Send a notification email to <z-user></z-user>?
+    {{#*inline "z-user"}}<strong>{{username}}</strong>{{#if email}} &#40;{{email}}&#41;{{/if}}{{/inline}}
```
I think it'd be cleaner here to just have the name, and not also the email address; we provided the email elsewhere and it's nicer for this to be short.

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -1,7 +1,29 @@
+<div class="email_field">
+    <p class="border-top">
+        <strong>From:</strong> Zulip administrator
+        <br/>
```
The actual "From" value will depend on the server's settings; let's just display the subject line here.

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -1,7 +1,29 @@
+    </p>
+    <div class="email-body">
+        <p>Dear {{username}},</p>
+        <p>Your Zulip account on {{realm_name}} has been deactivated, and you will no longer be able to log in.</p>
```
Let's provide the URL here; I think that's clearer.
```
Your Zulip account on {{realm_uri}} has been deactivated, ...
```

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -1,7 +1,29 @@
+    <div class="email-body">
+        <p>Dear {{username}},</p>
+        <p>Your Zulip account on {{realm_name}} has been deactivated, and you will no longer be able to log in.</p>
+        <p>The administrator for {{realm_name}} sent the following message:</p>
```
This can say "The administrators provided the following comment:"

timabbott commented on templates/zerver/emails/deactivate.source.html:
```diff
@@ -0,0 +1,16 @@
+<br/><br/>
+The administrator for {{realm_name}} sent the following message:
+
+<pre class="deactivated-user-text">{{ text }}</pre>
```
Why is this a `pre` block?  I would have thought a `blockquote` would be more appropriate for what we're doing (might require less CSS).

juliaBichler01 replied:
Because otherwise linebreaks for example are not recognized.

timabbott replied:
OK, we can use blockquote with just this line added:
```
 diff --git a/templates/zerver/emails/email.css b/templates/zerver/emails/email.css
index 4b554cea3d..ef78febd8e 100644
--- a/templates/zerver/emails/email.css
+++ b/templates/zerver/emails/email.css
@@ -332,6 +332,7 @@ a.button:hover {
     padding: 0 0 0 15px;
     margin: 0 0 20px;
     border-left: 5px solid #eee;
+    white-space: pre-line;
 }
 
 @media only screen and (max-width: 620px) {
```

timabbott commented on templates/zerver/help/deactivate-or-reactivate-a-user.md:
```diff
@@ -28,6 +28,33 @@ user's bots will also be deactivated. Lastly, the user will be unable to
+### Notify users of their deactivation
+
+If you want to notify the user of their deactivation,
+you can do so by sending them a email.
```
Zulip can optionally send the user an email notification that their account was deactivated.

timabbott commented on templates/zerver/help/deactivate-or-reactivate-a-user.md:
```diff
@@ -28,6 +28,33 @@ user's bots will also be deactivated. Lastly, the user will be unable to
+ 2. Click the **Deactivate** button to the right of the user account that you
+want to deactivate.
+
+ 3. Check the checkbox labeled **"Send a notification email to \[user]?"**.
```
Reading this, I'm reminded that we generally aim to not put details like the user's name in the checkbox.  Maybe call the checkbox "Notify this user by email" both here and in the actual UI?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8336,6 +8336,30 @@ paths:
+          schema:
+            type: string
+          example: Zoe
+          required: false
```
As noted elsewhere, we only need the text field, which perhaps we can call `deactivation_notification_comment`?

timabbott commented on zerver/views/users.py:
```diff
@@ -121,13 +130,26 @@ def deactivate_bot_backend(
+            to_emails=[target.email],
+            from_name="Zulip Administrator",
+            from_address=FromAddress.NOREPLY,
+            context={"text": email_text, "realm_name": realm_name, "username": username},
```
the realm name is just `user_profile.realm.name`; the username is `user_profile.full_name`.

timabbott commented:
@juliaBichler01 thanks for working on this!  I posted a batch of comments on a bunch of strings and some implementation details that could be improved.  I think we should be able to merge this fairly quickly once you've worked through the above.

Oh, one other note -- the API change to add the new parameter should come with bumping API_FEATURE_LEVEL and documentation in the API changelog.

timabbott commented on templates/zerver/help/deactivate-or-reactivate-a-user.md:
```diff
@@ -28,6 +28,33 @@ user's bots will also be deactivated. Lastly, the user will be unable to
+
+Below you can see how the email that the user receives will look like:
+
+<img src="/static/images/help/deactivate-user-email.png" alt="view-of-admin" width="600"/>
```
We should hold off on regenerating the screenshot until we've finalized all the text details, but we should leave off the address footer at the bottom of the email, since the fake address used in the development environment is kinda goofy :)

timabbott replied:
OK, we're ready to do this @juliaBichler01.

juliaBichler01 commented:
@timabbott thanks for the review, i implemented the requested changes :)

timabbott commented:
@juliaBichler01 thanks all your effort on this!  I pushed a bonus commit that fixes I believe all the remaining issues I identified with this PR; most of them seemed simpler to change than example:
* Use blockquote rather than <pre>.
* Send the email if the user requested it but left the textbox empty, with the custom comment not included.
* Tagged all the strings for translation.
* Changed the email sender to use the default "Zulip" sender.
* Set a limit of 2000 characters for the string.

Can you do another round of manual testing on this, and then add a commit that updates the Help Center documentation screenshot on top of this?  I can then squash and merge.

alya commented:
Yep, good questions! Those are both intentional. :)

1. Tim fixed the formatting [here](https://github.com/zulip/zulip/pull/20422#discussion_r803195562).
2. This change simplifies translation, and people's email clients should automatically linkify the URL, so we decided to just use plain text.

juliaBichler01 commented:
@alya Okay, good to know that this was intentional. 
Unfortunately, these changes don't really work, at least for me:
The formatting work correctly on http://localhost:9991/emails/, but not if i view those emails in my email client (Outlook):

On lokalhost:
![tmp](https://user-images.githubusercontent.com/74348920/153571896-0296c35c-3f8d-4fb3-a5eb-18a650a80d7e.png)

In Outlook:
![tmp2](https://user-images.githubusercontent.com/74348920/153571911-4295cd65-76d4-430d-b500-e1aac64f7065.png)

Also the realm uri is not clickable, even in outlook.

timabbott commented:
That's disappointing -- It's unfortunately only possible to determine what CSS features are supported by email clients by that sort of manual testing, so thanks for doing it!  I assume you remembered to run `inline-email-css`?  

I think the next thing we can try is using a `pre` tag (perhaps inside a blockquote) as you did in your original version, but adding CSS to declare a non-fixed-width font like our normal content in the email has, to try to create the blockquote look in a way that email clients will display correctly.

juliaBichler01 commented:
@timabbott Yes, i did run `inline-email-css`
I changed the way the comment and the links are displayed back to how i had it originally.
I've also already added the font for the pre-blocks a while ago (for reference: in `email.css` in line 24)

timabbott commented:
@juliaBichler01 can your resolve the merge conflicts on this one? I'll try to integrate it next week.

juliaBichler01 commented:
@timabbott i resolved the conflicts, it would be great if this could be integrated!
One of the checks is failing because of string capitalization, can that be accepted or should i try to change that?

timabbott commented on static/templates/confirm_dialog/confirm_deactivate_user.hbs:
```diff
@@ -18,3 +18,28 @@
+    <p class="border-top">
+        <strong>{{t "Subject" }}:</strong>
+        {{#tr}}
+        Notification of account deactivation on {organization_name}
```
We can use `realm_name`; I think you misinterpreted the linter error. Switched this back.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8643,6 +8643,20 @@ paths:
+            If not `""`, encodes custom text written by the administrator
+            to be included in the notification email.
+
+            **Changes**: New in Zulip 5.0 (feature level 115).
```
This feature level needs an update :)

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -1399,6 +1399,39 @@ def test_api(self) -> None:
+            result = self.client_delete(f"/json/users/{user.id}", dict())
+
+        self.assert_json_success(result)
+        self.assertFalse(n.called)
```
I rewrote this test to avoid mocking, and instead use the `outbox` to verify what actually happened -- check it out :)

timabbott commented:
This was super close, so I just took care of fixing the outstanding issues in order to merge this. In addition to what's noted above, it appears the `Fixes: #18943` was missing from the commit message, and 

@alya FYI -- I think the /help/ article included in this PR looks pretty good, but we may want to do another pass on it after doing a bit of testing of how the emails look when sent from a production environment.
